lib/Sema/CSDiag.cpp

```diff
@@ -767,11 +828,51 @@ static void diagnoseSubElementFailure(Expr *destExpr,
   }
 
   if (auto *ICE = dyn_cast<ImplicitConversionExpr>(immInfo.first))
-    if (isa<LoadExpr>(ICE->getSubExpr())) {
-      TC.diagnose(loc, diagID,
-                  "implicit conversion from '" +
-                      CS.getType(ICE->getSubExpr())->getString() + "' to '" +
-                      CS.getType(ICE)->getString() + "' requires a temporary")
+    if (auto *LE = dyn_cast<LoadExpr>(ICE->getSubExpr())) {
+      // Implicit conversions are most common with inout variables, and the default
+      // diagnostic is confusingly phrased, so emit a tailored diagnostic.
+      auto actualType = CS.getType(ICE->getSubExpr()), neededType = CS.getType(ICE);
+      if (diagID.ID == diag::cannot_pass_rvalue_inout_subelement.ID) {
+        TC.diagnose(loc, diag::cannot_pass_rvalue_inout_converted,
+                    actualType, neededType)
+          .highlight(ICE->getSourceRange());
+        
+        if (auto *DRE = dyn_cast<DeclRefExpr>(LE->getSubExpr())) {
+          if (auto * VD = dyn_cast_or_null<VarDecl>(DRE->getDecl())) {
```

Comment:
It's worth checking that this is a local variable; you wouldn't want to offer the fix-it for a global or a property.